src/expression/parse.js

```diff
@@ -916,17 +960,19 @@ export const createParse = /* #__PURE__ */ factory(name, dependencies, ({
 
     while (hasOwnProperty(operators, state.token)) {
       name = state.token
+      const source = tokenSource(state)
       fn = operators[name]
 
       getTokenSkipNewline(state)
 
       if (name === 'in' && state.token === '') {
         // end of expression -> this is the unit 'in' ('inch')
-        node = new OperatorNode('*', 'multiply', [node, new SymbolNode('in')], true)
+        // no source mapping because this * operator is not explicitly in the source expression
+        node = new OperatorNode('*', 'multiply', [node, withSources(new SymbolNode('in'), [source])], true)
       } else {
         // operator 'a to b' or 'a in b'
         params = [node, parseRange(state)]
-        node = new OperatorNode(name, fn, params)
+        node = withSources(new OperatorNode(name, fn, params), [source])
```

Comment:
The current implementation mutates Nodes by separately setting sources via `withSources` (or clearing them in `resolve`). I prefer an immutable solution instead. That could look something like: 

- introduce a new, optional constructor parameter `source` for all nodes, like `new OperatorNode(name, fn, params, [source])`
- extend the method `clone` on all nodes with an optional option `node.clone({ withSource: false })`
- in `resolve`, we can use `transform` and this clone method without copying the source.

What do you think?

Reply:
@josdejong my original implementation had constructors but we moved away from that in the discussion: https://github.com/josdejong/mathjs/discussions/2795#discussioncomment-3802631 . It'll be a big refactor so I want to double-check before making the change

If the goal is immutability, an optional constructor param is probably the best short-term solution. I like the idea of being able to opt out of sources in `clone()`

One thought:

```javascript
const node = new OperatorNode(name, fn, params, { sources });
node.clone({ sources: [] });
```

- Passing `{ sources }` as a dictionary of opts makes room for more options in the future without having a complicated list of ordered params
- Rather than `clone()` taking a true or false flag, we could allow users to optionally override the existing sources. It's more flexible and mirrors the syntax from the constructor. I think the idea of `clone()` allowing data to be overridden could be expanded on in the future to simplify this sort of code

```javascript
new OperatorNode(newName, oldNode.fn, oldNode.params)
```

But if that doesn't appeal to you, I'm perfectly fine with the syntax you outlined. Let me know which to implement and I'll get started 👍

Reply:
About `node.clone({ sources: [] })`: great idea, I like that! What should be the right default behavior for clone? I think copying the sources too?

About the constructor vs method `addSource`: having an immutable solution is really important to me. I hadn't realized that back then in  https://github.com/josdejong/mathjs/discussions/2795#discussioncomment-3802631. In both cases we _can_ implement an immutable solution. In case of a constructor we have that automatically, and in case of `addSource` we would have to return a cloned Node having the sources added. The first gives more "cluttered" constructors. The latter is bad for performance (in normal cases you create all nodes twice and throw away the first one). At this point I have a slight preference for the constructor. But maybe we should make that a bit more generic and extensible, like: `new OperatorNode(name, fn, params, meta: { sources })`, similar to your idea for `clone`. What are your thoughts about an (im)mutable API?

Reply:
In general, I like the idea of an immutable API for building the AST, it better fits the idea of Nodes as a data representation of the original expression.

I will say that the current constructors lean heavily on default params, and adding the `sources` param forces those params to be explicit in `parse.js`. It's especially bad for `AssignmentNode`:

![image](https://user-images.githubusercontent.com/114604111/229890511-cfdac4b6-73d4-46a2-a100-a7073a32da70.png)

where params can switch order:

```javascript
new AssignmentNode(symbol, value, null, { sources });
new AssignmentNode(symbol, index, value, { sources });
```

Long-term it might make more sense to move to a named-parameter style constructor, so order no longer matters:

```javascript
new AssignmentNode({ symbol, index, value, meta })
```

but that would be a pretty big breaking change

Reply:
Yes I totally agree! But indeed that would be a big refactor that we should discuss and implement separately.